### ISSUE TITLE
volksvga game doesnt work

### ISSUE THREAD

**Issue by stsp** [dosemu2/dosemu2#918]
**Describe the bug**
Music plays but no keyboard control.
Worked perfectly on 1.4,

**To Reproduce**
Just run.

**Attach the binaries**
[volksvga.exe.gz](https://github.com/stsp/dosemu2/files/3593072/volksvga.exe.gz)

**Comment by jschwartzenberg** [comment]
The keyboard seems to work fine here. I can get in-game and move with the numpad.

**Comment by stsp** [comment]
Argh, sorry!
It needs multiple files, and I only provided
an exe. Here it is:
[volsvga.tar.gz](https://github.com/stsp/dosemu2/files/3665437/volsvga.tar.gz)
Btw, this is the first program I wrote at school,
at a pascal language lessons.

**Comment by stsp** [comment]
When it asks you to configure sound card,
select Sound Blaster to reproduce the problem.
I am surprised it worked at all w/o all the resource
files...

**Comment by jschwartzenberg** [comment]
Impressive! Yes, I can reproduce with Sound Blaster, no sound allows the keyboard to work.

**Comment by stsp** [comment]
So needs a bisect.

**Comment by jschwartzenberg** [comment]
Already broken in pre8.

**Comment by jschwartzenberg** [comment]
I cannot reproduce this anymore over here. Maybe related to the version of DOS?

**Comment by stsp** [comment]
Indeed, but it was definitely not working!!!
What's going on?
Please check also the games in fdpp tracker,
as maybe they fixed themselves too. :(

**Comment by jschwartzenberg** [comment]
Maybe related to other system changes such as newer kernels? I switched notebooks though, I might still be able to reproduce on my other notebook.

**Comment by stsp** [comment]
Would be cool!

**Comment by jschwartzenberg** [comment]
Yeah, it still hangs there. I'll try to see if I can figure out the relevant difference.

**Comment by stsp** [comment]
See if carrier command is affected as well, #922

**Comment by jschwartzenberg** [comment]
I'm trying to break it now on my other system as well. Going back to the same dosemu revision did not break it. Now compiling fdpp.

Both systems run Ubuntu 18.04, but one notable difference is that one is on Linux 4.15 and the other on Linux 5.0.0.

**Comment by jschwartzenberg** [comment]
Matching fdpp & comcom32 doesn't break it either.

**Comment by stsp** [comment]
Neither fdpp nor comcom32 should matter
as I also tested with PC-DOS (which has neither
of those).

**Comment by stsp** [comment]
Is this still reproducible?

**Comment by jschwartzenberg** [comment]
Yes, with KVM it still hangs here.

**Comment by stsp** [comment]
OK, this appears to be the timer interrupt
storm. I have fixed a few related bugs, and
now it can be cured with something like this:
```
diff --git a/src/base/dev/pic/pic.c b/src/base/dev/pic/pic.c
index be3b768d..e56b457c 100644
--- a/src/base/dev/pic/pic.c
+++ b/src/base/dev/pic/pic.c
@@ -571,7 +571,7 @@ void run_irqs(void)
        /* don't allow HW interrupts in force trace mode */
        pic_activate();
        if (!isset_IF()) {
-               if (pic_pending())
+               if (pic_pending() && pic_sys_time < pic_dos_time + 50000)
                        set_VIP();
                return;                      /* exit if ints are disabled */
        }
```

This is a core2duo CPU, and it shows an
interesting thing: kvm can be much slower
than cpu-emu if you have high timer frequency.
The reason is that then VIP is set most of the
time, and running a single "iret" with a KVM
call is much more expensive than running
the same iret on cpu-emu.
Would it work if we always use cpu-emu when
VIP is set? Or any other work-arounds?
Of course we can resort to the usual "the reason
is now known, and this is all we need" rhetoric.
Bart?

**Comment by stsp** [comment]
Also I am pretty sure I've got that problem
too, when booted the debug fedora kernel.
So this is not only specific to an old CPUs.
So either switch to cpu-emu when VIP is
set (huge hack!), or maybe use the in-kernel
PIT for kvm...

**Comment by jschwartzenberg** [comment]
Impressive night work! Just tested that with the above patch the game goes further now :)

**Comment by bartoldeman** [comment]
If it's just about running a single `iret` that could be easily simulated in software without needing to go into KVM or invoking heavy duty cpuemu machinery.

in-kernel pit needs in-kernel pic which needs using IF instead of VIF so running with IOPL=3 inside KVM. It's all doable but a fair amount of work...

**Comment by stsp** [comment]
I haven't checked it to be exactly about
iret. It probably is about a few pops and
iret. :)
OK, in that case I'll cook something simple
for now, like a VIP hack.

**Comment by jschwartzenberg** [comment]
Any idea how QEMU deals with this?

**Comment by stsp** [comment]
I applied a slightly modified hack, could
you please test?
QEMU deals it by using the PIT in kvm.
And in case of tcg it deals with it the same
way we do with cpu-emu.

**Comment by jschwartzenberg** [comment]
Now it hangs here:
![image](https://user-images.githubusercontent.com/877112/76153919-dd933700-60d3-11ea-90ae-cb0abed800b7.png)

**Comment by stsp** [comment]
And now?
(force-pushed)

**Comment by jschwartzenberg** [comment]
Now it works!

**Comment by stsp** [comment]
Undone because of regressions.
